### ISSUE TITLE
Question about mrr calculation

### ISSUE THREAD

**Issue by zkid18** [allegro/allRank#17]
Hi everyone, thanks for releasing a great tool. 
I have a couple of question regarding mrr calucalation

The first one is why do you need cloning in this [place](https://github.com/allegro/allRank/blob/6167f9d9ca38e3a3e511f92003747756722a6ac5/allrank/models/metrics.py#L89):
```
y_true = y_true.clone()
y_pred = y_pred.clone()
```

Second question is about `padding_indicator`. 
Could you elaborate or provide any examples of using this parameters. 
I not quite into the ranking systems but I couldn't find any reference in the web yet.

**Comment by sadaharu-inugami** [comment]
Hello, thank you for your questions. 

1) Cloning is done to enable later in-place operations involving padding mask.

2) The `y_true` tensor contains relevance labels. The slate (sequence) lengths can be variable, so in a typical use case we assume an arbitrary length and assign the `padding_indicator` to any `y_true` indices exceeding the original slate length. Then, we can recognize these elements and not take them into account when calculating loss or metrics values. You can see how the sequences can be padded by investigating the behaviour of [load_libsvm_dataset_role](https://github.com/allegro/allRank/blob/6167f9d9ca38e3a3e511f92003747756722a6ac5/allrank/data/dataset_loading.py#L213) function. There, we pad the validation data to the maximum slate length and the train data to some arbitrary user-defined length, either sampling elements from the original sequence (if it is too long) or by adding padded elements (if it is not long enough). If we were to pad the train data to the maximum slate length, we would risk filling the GPU memory.

**Comment by zkid18** [comment]
Thanks for answer.

A bit confused about this one:
>  The slate (sequence) lengths can be variable, so in a typical use case we assume an arbitrary length and assign the padding_indicator to any y_true indices exceeding the original slate length

But according to docs `y_true` constrained by the slate size, isn't it?
```
    :param y_pred: predictions from the model, shape [batch_size, slate_length]
    :param y_true: ground truth labels, shape [batch_size, slate_length]
```

Say I have 100 items to rank and have a slate size equals 10. 
Why can't I simply take first/top `slate_szie` items `items[:slate_size]`

**Comment by PrzemekPobrotyn** [comment]
Hi!

in the MRR function (or any other metric or loss function) we assume inputs `y_pred` and `y_true` to be tensors, i.e. every slate in a batch to be already padded/truncated to the same length. If a given slate in a batch was padded to be longer, the corresponding `y_true` vector is padded to the same length with a `padding_inidcator` (which defaults to -1). Then, in the computation of a given metric or loss, these `padding_indicators` are used to determine which elements should not be taken into account when computing the value of the loss/metric. How to ignore these values depends on the details of a given loss/metric - sometimes it will involve setting certain values to 0, sometimes to `-inf`, it depends on the exact operations done to compute given metrics/loss (say, sorting, or softmax of ground truth, or something else). In the case of MRR you can see that we use the padding indicator to find the mask of elements to be ignored (in the `__apply_mask_and_get_true_sorted_by_preds` method) and then set `y_pred` elements corresponding to the mask to `-inf` (because we want them to be last elements of the tensor after sorting) and the corresponding `y_true` elements to 0 (cause we don't want them to contribute to the metric value).

If all slates in your batch are of the same length, then none of the above operations will actually take place. But if your batch consists of slates or varying lengths, one first pads/truncates them to be the same length and then does the masking logic as explained above in order to ignore the padded values in metrics/loss value computations.

I hope this answers your question, please let me know if anything is still not clear.
